pkg/controller/master/upgrade/upgrade_repo.go

```diff
@@ -303,6 +309,74 @@ func (r *UpgradeRepo) createVM(image *harvesterv1.VirtualMachineImage) (*kubevir
 	return r.h.vmClient.Create(&vm)
 }
 
+func (r *UpgradeRepo) deleteVM() error {
+	vmName := r.getVMName()
+
+	vm, err := r.h.vmCache.Get(upgradeNamespace, vmName)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return r.deleteImage("")
+		}
+		return err
+	}
+
+	if pvcName, ok := vm.Annotations[util.RemovedPVCsAnnotationKey]; ok {
+		if err := r.deleteImage(pvcName); err != nil {
+			return err
+		}
+	}
+
+	logrus.Infof("Delete upgrade repo VM %s/%s", vm.Namespace, vm.Name)
+	return r.h.vmClient.Delete(vm.Namespace, vm.Name, &metav1.DeleteOptions{})
+}
+
+// deleteImage deletes a repo image
+// (1) If a PVC based on the image exists, it adds the PVC as the OwnerReference to the repo image.
+// Once the repo VM is deleted and VM's PVC is deleted, the repo image will be garbage collected.
+// (2) If there is no PVCs based on the repo image. We delete the image directly.
+func (r *UpgradeRepo) deleteImage(pvcName string) error {
```

Comment:
Maybe we can keep the image as the upgrade may fail or they can delete it manually?

Reply:
I'm good to keeping it.
Just the original issue states the image needs to be deleted as well. @TachunLin any takes?

Reply:
Checked the upgrade image can't be used on for normal re-upgrade user behavior. 
I would suggest to delete the image to save some space and clean up the Volumes page.

Reply:
@guangbochen As David pointed out,  the kept image can't be used by users directly. With normal UX, the user still needs to trigger an upgrade from a remote ISO. Thus the image needs to be downloaded again.